utils/threadpool/fair_threadpool.cpp

```diff
@@ -0,0 +1,295 @@
+                               uint ID)
+ : _stop(false), weightPerRun(targetWeightPerRun), id(ID), blockedThreads(0), extraThreads(0), stopExtra(true)
+{
+  boost::thread* newThread;
+  for (uint32_t i = 0; i < highThreads + midThreads +lowThreads; ++i)
+  {
+    newThread = threads.create_thread(ThreadHelper(this, HIGH));
+    newThread->detach();
+  }
+  cout << "FairThreadPool started " << highThreads + midThreads + lowThreads << " thread/-s.\n";
+  defaultThreadCounts = threadCounts = highThreads + midThreads + lowThreads;
+}
+
+FairThreadPool::~FairThreadPool()
+{
+  stop();
+}
+
+void FairThreadPool::addJob(const Job& job, bool useLock)
+{
+  boost::thread* newThread;
+  boost::mutex::scoped_lock lk(mutex, boost::defer_lock_t());
+
+  if (useLock)
+    lk.lock();
+
+  // Create any missing threads
+  if (defaultThreadCounts != threadCounts)
+  {
+    newThread = threads.create_thread(ThreadHelper(this, HIGH));
+    newThread->detach();
+    ++threadCounts;
+  }
+
+  // If some threads have blocked (because of output queue full)
+  // Temporarily add some extra worker threads to make up for the blocked threads.
+  if (blockedThreads > extraThreads)
+  {
+    stopExtra = false;
+    newThread = threads.create_thread(ThreadHelper(this, EXTRA));
+    newThread->detach();
+    extraThreads++;
+  }
+  else if (blockedThreads == 0)
+  {
+    // Release the temporary threads -- some threads have become unblocked.
+    stopExtra = true;
+  }
+
+  auto jobsListMapIter = txn2JobsListMap_.find(job.txnIdx_);
+  if (jobsListMapIter == txn2JobsListMap_.end())
+  {
+    ThreadPoolJobsList* jobsList = new ThreadPoolJobsList;
+    jobsList->push_back(job);
+    txn2JobsListMap_[job.txnIdx_] = jobsList;
+    WeightT currentTopWeight = weightedTxnsQueue_.empty() ? 0 : weightedTxnsQueue_.top().first;
+    weightedTxnsQueue_.push({currentTopWeight, job.txnIdx_});
+  }
+  else
+  {
+    jobsListMapIter->second->push_back(job);
+  }
+
+  if (useLock)
+    newJob.notify_one();
+}
+
+void FairThreadPool::removeJobs(uint32_t id)
+{
+  boost::mutex::scoped_lock lk(mutex);
+
+  for (auto& txnJobsMapPair : txn2JobsListMap_)
+  {
+    ThreadPoolJobsList* txnJobsList = txnJobsMapPair.second;
+    auto job = txnJobsList->begin();
+    while (job != txnJobsList->end())
+    {
+      if (job->id_ == id)
+      {
+        job = txnJobsList->erase(job); // update the job iter
+        if (txnJobsList->empty())
+        {
+          txn2JobsListMap_.erase(txnJobsMapPair.first);
+          delete txnJobsList;
+          break;
+          // There is no clean-up for PQ. It will happen later in threadFcn
+        }
+        continue; // go-on skiping job iter increment
+      }
+      ++job;
+    }
+  }
+}
+
+void FairThreadPool::threadFcn(const Priority preferredQueue)
+{
+  if (preferredQueue == EXTRA)
+    utils::setThreadName("Extra");
+  else
+    utils::setThreadName("Idle");
+  RunListT runList; // This is a vector to allow to grab multiple jobs
```

Comment:
Just wondering, why do we use a vector, since it's for a single job?

Reply:
In my next patch I will benchmark FairThreadPool grabbing multiply jobs to reduce TP scheduling algo overhead so this jobs collection will be either a vector or a list.